### ISSUE TITLE
False positives from highsecure.ru fuzzy storage server.

### ISSUE THREAD

**Issue by moisseev** [vstakhov/rspamd#366]
I'm having false positives from highsecure.ru:11335 for some subscribe.ru list messages.
Encrypted message sample:
http://my-files.ru/5ub9j4

**Comment by vstakhov** [comment]
I think I should just create a whitelist of domains for fuzzy learning.

**Comment by moisseev** [comment]
Did you mean you can't remove that hash from the storage because of it will be relearned again due to learning procedure you are using? Maybe you can delete it from FUZZY_DENIED and add to FUZZY_WHITE? Or suggest any workaround?

**Comment by vstakhov** [comment]
I can remove it. I just want to be able to skip fuzzy checks/learns for the case of some domains.

**Comment by vstakhov** [comment]
Well, my current thought is to use `condition` for symbols properly. The idea is the following: each rule in rspamd can have `condition` - a function in lua that returns boolean value specifying when a rule should be executed. This allows to extend whitelisting functionality for any symbol in the following way:

~~~lua
rspamd_config:add_condition('FUZZY_DENY', function(task)
  if some_map:find_key(task:get_from()) then return false end
  return true
end)
~~~

This is a rough example of what I'm talking about.

**Comment by moisseev** [comment]
Why not remove it now?

It interesting idea, but fake domains will be whitelisted as well, I'm afraid.

**Comment by vstakhov** [comment]
I cannot download it, I'm afraid. Your link leads me to some strange page with tonns of ads and afterwards it got blocked by my corporate firewall.

**Comment by moisseev** [comment]
I've sent copy" to your email.

**Comment by vstakhov** [comment]
I've removed it, thank you!

**Comment by moisseev** [comment]
I can't confirm that. Still getting FP.
```
2015-09-28 20:43:45 #85979(controller) <ed811d>; task; fuzzy_io_callback: <20150917161904.hui.10637855@10637855.news.citycat.subscribe>, found fuzzy hash with weight: 0.99, in list: FUZZY_DENIED:1
```
```
fuzzy_check {
    # Include dynamic conf for the rule
    .include(try=true,priority=1) "${DBDIR}/dynamic/fuzzy_check.conf"
    min_bytes = 300;
    rule {
        servers = "rspamd.com:11335";
        symbol = "FUZZY_UNKNOWN";
        mime_types = ["application/*"];
        max_score = 20.0;
        read_only = yes;
        skip_unknown = yes;
        fuzzy_map = {
            FUZZY_DENIED {
                max_score = 20.0;
                flag = 1;
            }
...
```

**Comment by moisseev** [comment]
It's finally gone. But why it took several days after you removed it? Is there a cache somewhere?

**Comment by vstakhov** [comment]
Are you still able to find other hashes? I've enabled optional encryption yesterday, so I'm curious if unencrypted requests are still processed.

**Comment by moisseev** [comment]
I think I should be able to find hashes. I've got the encryption_key )). I'll test it if you give me some message samples.

**Comment by moisseev** [comment]
You are right. Despite I've configured encryption_key I haven't seen any matched hash since. But today I've got several matches and the false positives as well.

**Comment by vstakhov** [comment]
I can hardly believe that encryption has any influence on finding hashes. With `1.0.4` rspamd now tries to retransmit missing requests, that should improve the quality of matching for many cases.

**Comment by moisseev** [comment]
I found the problem. I checked wrong message. Sorry.
I've sent three more samples to you in private. Would you remove those hashes too?

**Comment by nahuelange** [comment]
Hi,
Anyway, highsecure seems to not work ATM:

```2015-11-17 06:28:33 #11558(normal) fuzzy_io_callback: got error on IO with server highsecure.ru:11335, 110, Connection timed out ```

**Comment by vstakhov** [comment]
It was too busy for several days because of sqilte issues. Now, I've resolved some of them and it should work as intended.

**Comment by moisseev** [comment]
@vstakhov Would you please remove `subscribe.ru` logo image hash from rspamd.com fuzzy storage?
It is in this part
~~~
^M
--=felis-related=20160317225554=846635^M
Content-Type: image/png;^M
  name="=?utf-8?Q?logo=2Epng?="^M
Content-Transfer-Encoding: base64^M
Content-Location: http://subscribe.ru/img/new/index/logo.png^M
^M
~~~
of the message sample:
https://paste.rspamd.com/qy42/raw

Since rspamc return just scores `[1: 1.00 / 1.00, 1: 1.00 / 1.00]` I had to check each message part one by one to find it out. Is there some easier way to find out what message part is matching a fuzzy hash?

**Comment by vstakhov** [comment]
On 18/03/2016 07:03, Alexander Moisseev wrote:
> @vstakhov <https://github.com/vstakhov> Would you please remove
> |subscribe.ru| logo image hash from rspamd.com fuzzy storage?
> It is in this part
> 
> |^M --=felis-related=20160317225554=846635^M Content-Type: image/png;^M
> name="=?utf-8?Q?logo=2Epng?="^M Content-Transfer-Encoding: base64^M
> Content-Location: http://subscribe.ru/img/new/index/logo.png^M ^M |
> 
> of the message sample:
> https://paste.rspamd.com/qy42/raw

Great, I've removed that hashes and added subscribe.ru to fuzzy learn
skip-list.

> Since rspamc return just scores |[1: 1.00 / 1.00, 1: 1.00 / 1.00]| I had
> to check each message part one by one to find it out. Is there some
> easier way to find out what message part is matching a fuzzy hash?

That is desired feature. Could you open a separate issue please just for
my records?

-- 
Vsevolod Stakhov

**Comment by moisseev** [comment]
Another subscribe.ru image.

https://paste.rspamd.com/Z1FA/raw

An HDTV advertisement at the top :
~~~
--=felis-related=20160324180718=328195
Content-Type: image/png;
  name="=?utf-8?Q?=5F13064=2F524009=2Fimg=2Epng?="
Content-Transfer-Encoding: base64
Content-Location: http://image.subscribe.ru/advert/13064/524009/img.png
Content-ID: <advert/13064/524009/img.png>
~~~

**Comment by vstakhov** [comment]
I've removed that hash. In fact, all `subscribe.ru` messages are filtered in my configuration:

```
return function(task)
        local skip_domains = {
                'subscribe.ru',
                'google.com',
        }

        local from = task:get_from()

        if from and from[1] and from[1]['addr'] then
                for i,d in ipairs(skip_domains) do
                        if string.find(from[1]['addr'], d) then
                                return false
                        end
                end
        end

        return true
end
EOD
```

**Comment by moisseev** [comment]
Thanks!
I think the hashes was learned before you filtered out that domains and maybe some of them are still in the storage.

Spam messages can potentially have the same images as ham messages. So fuzzy storage pollution is possible. I have no clue how to deal with it, but luckily I haven't seen any yet.

**Comment by moisseev** [comment]
Another hash
https://paste.rspamd.com/kwt5/raw

**Comment by moisseev** [comment]
Please remove the hash of the Avast footer.
~~~
# rspamc avast-cut.eml | fgrep FUZZY_DENIED
Symbol: FUZZY_DENIED (11.99)[1:c2d2369fd7:1.00]
~~~

~~~
MIME-Version: 1.0
Content-Type: multipart/mixed;
 boundary="----------04F0E71601CE9C73F"
X-Antivirus: avast! (VPS 160412-1, 12.04.2016), Outbound message
X-Antivirus-Status: Clean

------------04F0E71601CE9C73F
Content-Type: text/plain; charset=windows-1251
Content-Transfer-Encoding: quoted-printable


---
=DD=F2=EE =F1=EE=EE=E1=F9=E5=ED=E8=E5 =EF=F0=EE=E2=E5=F0=E5=ED=EE =ED=E0 =
=E2=E8=F0=F3=F1=FB =E0=ED=F2=E8=E2=E8=F0=F3=F1=EE=EC Avast.
https://www.avast.com/antivirus

~~~
